src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp

```diff
@@ -0,0 +1,147 @@
+    }
+  }
+  return result;
+}
+
+// Compute the indicator matrix, as given by Eq. 28 of Dumbser2007.
+//
+// The indicator matrix is computed by tensor product, using the tensor-product
+// nature of the Legendre basis in N dimensions.
+//
+// Note that we make a change (correction?) to the reference. In each direction
+// of the tensor product, we sum from 0 derivatives to the maximum number,
+// instead of from 1 to the maximum as in the paper. With this change, the
+// tensor product includes terms that have derivatives in only a subset of the
+// spatial dimensions. We must however subtract off the term which has no
+// derivatives in any dimension.
+template <size_t VolumeDim>
+Matrix compute_indicator_matrix(const Mesh<VolumeDim>& mesh) noexcept {
+  Matrix result(mesh.number_of_grid_points(), mesh.number_of_grid_points());
+  for (IndexIterator<VolumeDim> m(mesh.extents()); m; ++m) {
+    for (IndexIterator<VolumeDim> n(mesh.extents()); n; ++n) {
+      // Compute each indicator matrix term by tensor-product of 1D Legendre
+      // polynomial math
+      auto& result_mn = result(m.collapsed_index(), n.collapsed_index());
+      result_mn =
+          compute_sum_of_legendre_derivs(mesh.extents(0), m()[0], n()[0]);
+      for (size_t dim = 1; dim < VolumeDim; ++dim) {
+        result_mn *= compute_sum_of_legendre_derivs(mesh.extents(dim), m()[dim],
+                                                    n()[dim]);
+      }
+    }
+    // Subtract the tensor-product term that has no derivatives
+    double term_with_no_derivs =
+        Spectral::compute_basis_function_normalization_square<
+            Spectral::Basis::Legendre>(m()[0]);
+    for (size_t dim = 1; dim < VolumeDim; ++dim) {
+      term_with_no_derivs *=
+          Spectral::compute_basis_function_normalization_square<
+              Spectral::Basis::Legendre>(m()[dim]);
+    }
+    result(m.collapsed_index(), m.collapsed_index()) -= term_with_no_derivs;
+  }
+  return result;
+}
+
+}  // namespace
+
+namespace SlopeLimiters {
+namespace Weno_detail {
+
+template <size_t VolumeDim>
+double oscillation_indicator(const DataVector& data,
+                             const Mesh<VolumeDim>& mesh) noexcept {
+  ASSERT(mesh.basis() == make_array<VolumeDim>(Spectral::Basis::Legendre),
+         "No implementation for mesh: " << mesh);
+  ASSERT(mesh.quadrature() ==
+             make_array<VolumeDim>(Spectral::Quadrature::GaussLobatto),
+         "No implementation for mesh: " << mesh);
+
+  // TODO(): Generalize the caching infrastructure to handle multiple different
+  // matrices. This will be necessary before (efficiently) using domains where
+  // different elements may have different meshes, and different matrices.
+  static const Matrix indicator_matrix = compute_indicator_matrix(mesh);
+
+  const ModalVector coeffs = to_modal_coefficients(data, mesh);
+
+  double result = 0.;
+  for (size_t m = 0; m < mesh.number_of_grid_points(); ++m) {
```

Comment:
I do agree this could start at 1; "physically", the mean doesn't contribute any oscillation, so the 0'th spectral coeff shouldn't play into the indicator.

I've changed the code to compute an (N-1)^2 indicator matrix and contract it with (N-1) modal coeffs (the numerical results are unchanged within the precision). This is less elegant, but may provide a tiny performance benefit.